### PR TITLE
Improve write buffer performance

### PR DESCRIPTION
**Pull request by joaander**
## Description

* Write data chunks to the file first in, first out.
* Flush the data chunk write buffer as needed when writing chunks, but buffer index entries indefinitely.
* Remove all implicit calls to `gsd_flush` except in `gsd_close`.
* Remove the no longer relevant `index_entries_to_buffer`.
* Change the default buffer to 1 MB.

These changes improve the streaming write performance when writing large files because there is no longer a need to sync when flushing the write buffer. The caller now gets explicit control of when the index buffer is written to disk. As a consequence, files open in read/write mode can no longer read chunks until after an explicit call to `gsd_flush`.

## Motivation and Context

During a review of the GSD code, I discovered many problems:
* `fsync` does not behave the same on macOS as it does on Linux (see the comment for b6f94b6)
* There was a missing sync after writing the index.
* The frame_index/buffer_index code was needlessly complicated.

I fixed these problems. Unfortunately, this drastically reduced performance on macOS and Frontier. At the same time, users have found that the default buffer size of 64 MB is too large and they often call flush on every frame.

`flush` is now *more necessary* than it was before because it is never called implicitly (except on close). I plan to solve the user experience problem in HOOMD with an automatic flush based on a timer. Flushing based on the number of frames or bytes written kills performance (especially on macOS and Frontier), but a timer based solution will naturally spread the (relatively) fixed cost flushes evenly and can provide users with consistent behavior regardless of all other settings (e.g., new frames will appear in the file every 10 seconds).

## How Has This Been Tested?

New unit tests cover many corner cases in the new implementation.

Here is benchmark-write performance as a function of buffer size on a variety of systems:

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/gsd/blob/trunk-patch/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**GSD Contributor Agreement**](https://github.com/glotzerlab/gsd/blob/trunk-patch/ContributorAgreement.md).
- [x] My name is on the list of contributors (`doc/credits.rst`) in the pull request source branch.
- [x] I have added a change log entry to ``CHANGELOG.rst``.

### CODE REVIEW

File: gsd/gsd.c [hunk 1]
```diff
@@ -68,13 +68,7 @@ enum
...
 enum
     {
-    GSD_DEFAULT_MAXIMUM_WRITE_BUFFER_SIZE = 64 * 1024 * 1024
-    };
-
-/// Default number of index entries to buffer
-enum
-    {
-    GSD_DEFAULT_INDEX_ENTRIES_TO_BUFFER = 256 * 1024
+    GSD_DEFAULT_MAXIMUM_WRITE_BUFFER_SIZE = 1024 * 1024
```
Comment (janbridley) [inline gsd/gsd.c]: Would be nice to have a code comment with units here - MB/MiB?

File: gsd/test/test_fl.py [hunk 2]
```diff
@@ -1097,13 +1104,98 @@ def test_read_write(tmp_path, mode):
...
+        assert f.nframes == nframes
+
         for i in range(nframes):
             data1 = f.read_chunk(frame=i, name='data1')
             data10 = f.read_chunk(frame=i, name='data10')
             assert data1[0] == i
             assert data10[0] == i * 10
 
 
+@pytest.mark.parametrize('mode', ['w', 'x', 'a', 'r+'])
```
Comment (janbridley) [inline gsd/test/test_fl.py:1118]:
```suggestion
@pytest.mark.parametrize('mode', ['w', 'x', 'a', 'r', 'r+'])
```

File: gsd/test/test_fl.py [hunk 3]
```diff
@@ -1211,7 +1315,34 @@ def test_pending_index_entries(tmp_path):
...
+        for i in range(N_ENTRIES):
+            data = numpy.array([i], dtype=numpy.int64)
+            f.write_chunk(name='data', data=data)
+            f.end_frame()
+            if flush and (i & 0xF == 0):
+                f.flush()
+
+        f.flush()
+        for i in range(N_ENTRIES):
+            expected_data = numpy.array([i], dtype=numpy.int64)
```
Comment (janbridley) [inline gsd/test/test_fl.py]: I'm not sure if this is valid/the correct thing to test, but it could be nice to validate the expansion for different sized inputs.

```suggestion
            data = numpy.array([i] * i, dtype=numpy.int64)
            f.write_chunk(name='data', data=data)
            f.end_frame()
            if flush and (i & 0xF == 0):
                f.flush()

        f.flush()
        for i in range(N_ENTRIES):
            expected_data = numpy.array([i] * i, dtype=numpy.int64)
```

**Comment by janbridley** [comment]
I'm not super familiar with this codebase but it looks good to me -  I added a few small comments and one change to the tests, but its great overall.

File: gsd/test/test_fl.py (hunk 3 above)
Comment (joaander) [reply gsd/test/test_fl.py]: Sure.

File: gsd/gsd.c (hunk 1 above)
Comment (joaander) [reply gsd/gsd.c]: I clarified the units of this value.

File: gsd/gsd.h [hunk 4]
```diff
@@ -462,7 +459,10 @@ extern "C"
         @pre *handle* was opened by gsd_open().
 
         @post The current frame counter is increased by 1.
-        @post Flush the write buffer if it has overflowed. See gsd_flush().
+
+        @note Starting with GSD 4.0, gsd_end_frame() does NOT automatically flush buffered frames
+        to the filesystem. Callers must manually call gsd_flush() or gsd_close() to commit the
```
Comment (DomFijan) [inline gsd/gsd.h:466]: Perhaps it would make sense to document that when one calls gsd_close() calling gsd_flush() is not necessary in gsd_close() docs.

Comment (joaander) [reply gsd/gsd.h:466]: Sure thing.
